presto-main/src/test/java/io/prestosql/sql/planner/TestUnwrapCastInComparison.java

```diff
@@ -0,0 +1,516 @@
+                anyTree(
+                        filter("NOT (A IS NULL) OR NULL",
+                                values("A"))));
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a <> DOUBLE '1.9'",
+                anyTree(
+                        filter("NOT (A IS NULL) OR NULL",
+                                values("A"))));
+
+        // top of range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a <> DOUBLE '32767'",
+                anyTree(
+                        filter("A <> SMALLINT '32767'",
+                                values("A"))));
+
+        // above range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a <> DOUBLE '32768'",
+                anyTree(
+                        filter("NOT (A IS NULL) OR NULL",
+                                values("A"))));
+
+        // bottom of range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a <> DOUBLE '-32768'",
+                anyTree(
+                        filter("A <> SMALLINT '-32768'",
+                                values("A"))));
+
+        // below range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a <> DOUBLE '-32769'",
+                anyTree(
+                        filter("NOT (A IS NULL) OR NULL",
+                                values("A"))));
+    }
+
+    @Test
+    public void testLessThan()
+    {
+        // representable
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a < DOUBLE '1'",
+                anyTree(
+                        filter("A < SMALLINT '1'",
+                                values("A"))));
+
+        // non-representable
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a < DOUBLE '1.1'",
+                anyTree(
+                        filter("A <= SMALLINT '1'",
+                                values("A"))));
+
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a < DOUBLE '1.9'",
+                anyTree(
+                        filter("A < SMALLINT '2'",
+                                values("A"))));
+
+        // top of range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a < DOUBLE '32767'",
+                anyTree(
+                        filter("A <> SMALLINT '32767'",
+                                values("A"))));
+
+        // above range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a < DOUBLE '32768'",
+                anyTree(
+                        filter("NOT (A IS NULL) OR NULL",
+                                values("A"))));
+
+        // bottom of range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a < DOUBLE '-32768'",
```

Comment:
You should test the value one up from this.  Basically, for all of the comparison operators like this, I'd like to see one value that doesn't translate, and then the false/true values